third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp

```diff
@@ -763,50 +767,100 @@ class ScaledBlockedToScaledMFMAF8F6F4 final
     auto mfmaEnc = ttg::AMDMfmaEncodingAttr::get(
         ctx, /*versionMajor=*/mfmaVersion, /*versionMinor=*/0, warpsPerTile,
         /*instrShape=*/mDim, nDim, /*isTransposed=*/true, ctaLayout);
+    auto warpOrder = mfmaEnc.getWarpOrder();
 
     auto newRetType =
         RankedTensorType::get(oldShape, oldRetType.getElementType(), mfmaEnc);
 
     auto newAcc = rewriter.create<ttg::ConvertLayoutOp>(
         dotOp.getC().getLoc(), newRetType, dotOp.getC());
 
+    StringAttr kRegister = StringAttr::get(ctx, "register");
+    StringAttr kLane = StringAttr::get(ctx, "lane");
+    StringAttr kWarp = StringAttr::get(ctx, "warp");
+    StringAttr kBlock = StringAttr::get(ctx, "block");
+
+    auto order = ttg::getMatrixOrder(rank, /*rowMajor=*/true);
+    auto standardOutDims = standardOutDimNames(ctx, rank);
+
     // For the mfma_scale_f32_*_f8f6f4 instructions, each thread consumes 32
     // elements. But since two fp4 elements are packed into one int8, the
     // kWidth is 16 for fp4.
-    unsigned kWidth = kBase;
-    auto newAEncoding = DotOperandEncodingAttr::get(
-        ctx, /*idx=*/0, newRetType.getEncoding(),
-        aElemType == ScaleDotElemType::E2M1 ? kWidth / 2 : kWidth);
+    const unsigned kWidth = kBase;
+    using basisT = std::vector<std::vector<int32_t>>;
+
+    // For mxfp4, we use dot layout directly. Mxfp8 is not covered by dot
+    // layout, so we need to manually create linear layout for it.
+    auto createOperandLinearLayout = [&](int idx, ScaleDotElemType elemType,
+                                         llvm::ArrayRef<int64_t> shape) {
+      if (elemType == ScaleDotElemType::E2M1) {
+        auto newEncoding = DotOperandEncodingAttr::get(
+            ctx, idx, newRetType.getEncoding(), kWidth / 2);
+        return newEncoding.toLinearLayout(shape);
+      } else {
```

Comment:
Style nit: no need for `} else {` given we have early return in the above so that we can save one level of indent for the following.